formats/notion/copy-files.ts

```diff
@@ -0,0 +1,119 @@
+import { readFile } from 'fs/promises';
+import { ImportResult } from 'main';
+import moment from 'moment';
+import { App, normalizePath } from 'obsidian';
+import { assembleParentIds, parseDate } from './notion-utils';
+import { getParentFolder } from '../../util';
+
+export async function copyFiles({
+	idsToFileInfo,
+	pathsToAttachmentInfo,
+	attachmentFolderPath,
+	targetFolderPath,
+	app,
+	results,
+}: {
+	idsToFileInfo: Record<string, NotionFileInfo>;
+	pathsToAttachmentInfo: Record<string, NotionAttachmentInfo>;
+	attachmentFolderPath: string;
+	targetFolderPath: string;
+	app: App;
+	results: ImportResult;
+}) {
+	const createdFolders = new Set<string>();
+
+	await Promise.all(
```

Comment:
Wow, that saves so much headache. I will switch to sequential processing.

Reply:
Actually after switching to sequential processing, the import process takes much longer. Even if the internal fs serializes I think keeping parallel processing speeds it up significantly. What do you think?

Reply:
> Hmm interesting. How much of a difference is it? Do you have any intuition as to why that is? Could you run a performance profile for before and after?
> 
> I'm currently thinking it is possibly due to the parallelization of the `fs.readFile` call for the attachments.
> 
> If that's the case we can look into parallelization again, though I'd need to think about how to keep the code comprehensible 🤣

Let me do that. For importing 1000 notes it used to take 2 seconds now it takes more than 10. I think `readFile` is a big part, but am not sure. I'll profile tomorrow

Reply:
Here's the profile for sequential processing...`json` isn't supported so I zipped it. As a side note it could have been some plugins I had enabled that responded to `metadataCache` events adding a lot of the time. Still though something to think about.

[Profile-20230802T093420.json.zip](https://github.com/obsidianmd/obsidian-importer/files/12241925/Profile-20230802T093420.json.zip)

Reply:
Thank you! I've gone back to putting more thought into this problem and I think I am inclined to forego parallelization for the following reasons:

- Memory limits. If we work in parallel, this could lead to out-of-memory cases, especially for concurrently downloading multiple attachments. This is a much bigger problem on mobile devices, if/when we want to support them in the future.
- One-time use. Because the use case of the importer isn't a day-to-day thing, I think most people will be ok to wait a few more seconds for their import. I'd want to make sure we don't do anything obviously inefficient, but I think maximum parallelization isn't something we need to achieve.

What do you think?

Reply:
I agree. Given that the importer already needs to load a lot of data in memory it makes sense not to read/write files in parallel.